framework/Optimizers/RavenSampled.py

```diff
@@ -0,0 +1,566 @@
+
+  def amIreadyToProvideAnInput(self):
+    """
+      This is a method that should be called from any user of the optimizer before requiring the generation of a new input.
+      This method act as a "traffic light" for generating a new input.
+      Reason for not being ready could be for example: exceeding number of model evaluation, convergence criteria met, etc.
+      @ In, None
+      @ Out, ready, bool, indicating the readiness of the optimizer to generate a new input.
+    """
+    # if any trajectories are still active, we're ready to provide an input
+    ready = Optimizer.amIreadyToProvideAnInput(self)
+    # we're not ready yet if we don't have anything in queue
+    ready = ready and len(self._submissionQueue)
+    return ready
+
+  def localGenerateInput(self, model, inp):
+    """
+      Provides the next sample to take.
+      After this method is called, the self.inputInfo should be ready to be sent
+      to the model
+      @ In, model, model instance, an instance of a model
+      @ In, inp, list, a list of the original needed inputs for the model (e.g. list of files, etc.)
+      @ Out, None
+    """
+    # get point from stack
+    point, info = self._submissionQueue.popleft()
+    point = self.denormalizeData(point)
+    # assign a tracking prefix
+    prefix = self.inputInfo['prefix']
+    # register the point tracking information
+    self._registerSample(prefix, info)
+    # build the point in the way the Sampler expects
+    for var in self.toBeSampled: #, val in point.items():
+      val = point[var]
+      self.values[var] = val # TODO should be np.atleast_1d?
+      ptProb = self.distDict[var].pdf(val)
+      # sampler-required meta information # TODO should we not require this?
+      self.inputInfo['ProbabilityWeight-{}'.format(var)] = ptProb
+      self.inputInfo['SampledVarsPb'][var] = ptProb
+    self.inputInfo['ProbabilityWeight'] = 1 # TODO assume all weight 1? Not well-distributed samples
+    self.inputInfo['PointProbability'] = np.prod([x for x in self.inputInfo['SampledVarsPb'].values()])
+    self.inputInfo['SamplerType'] = self.type
+
+  def localFinalizeActualSampling(self, job, model, inp):
+    """
+      Runs after each sample is collected from the JobHandler.
+      @ In, job, Runner instance, job runner entity
+      @ In, model, Model instance, RAVEN model that was run
+      @ In, inp, list, generated inputs for run
+      @ Out, None
+    """
+    Optimizer.localFinalizeActualSampling(self, job, model, inp)
+    # TODO should this be an Optimizer class action instead of Sampled?
+    # collect finished job
+    prefix = job.getMetadata()['prefix']
+    # If we're not looking for the prefix, don't bother with using it
+    ## this usually happens if we've cancelled the run but it's already done
+    if not self.stillLookingForPrefix(prefix):
+      return
+    if job.getReturnCode() != 0: # TODO shouldn't this be "if job failed"?
```

Comment:
we should not need to "finalizeActualSampling" if a run failed... (The step calls the handle failed run in case)... Right?

Reply:
removed, cool